compiler/rustc_codegen_llvm/src/intrinsic.rs

```diff
@@ -532,6 +534,22 @@ impl<'ll, 'tcx> IntrinsicCallBuilderMethods<'tcx> for Builder<'_, 'll, 'tcx> {
                 return Ok(());
             }
 
+            sym::dynamic_shared_memory => {
+                let global = self.declare_global_in_addrspace(
+                    "dynamic_shared_memory",
+                    self.type_array(self.type_i8(), 0),
+                    AddressSpace::SHARED,
+                );
+                let ty::RawPtr(inner_ty, _) = result.layout.ty.kind() else { unreachable!() };
+                let alignment = self.align_of(*inner_ty).bytes() as u32;
+                unsafe {
+                    if alignment > llvm::LLVMGetAlignment(global) {
+                        llvm::LLVMSetAlignment(global, alignment);
+                    }
+                }
```

Comment:
What does this do when there are multiple translation units? It seems like each would declare its own version of the global, potentially with a different alignment?

Reply:
> All of them return the same pointer.

By which magic does that happen? That's not usually how LLVM globals work, AFAIK.

> I assume the linker ensures the maximum alignment is taken and the runtime will ensure that it’s honored.

That would be rather surprising to me; usually a global can only be declared in one translation unit and then must be imported in all the others. Your current implementation might end up with a separate symbol for each TU, I am not sure.

> (Though note that multiple translation units are currently not really supported in the backends, they both rely on IR linking.)

How does that work in rustc where each crate is a separate TU?

Reply:
Usually an alignment attribute on an extern global encodes an assumption -- it's a promise to LLVM that someone else has aligned the global sufficiently.

It looks like this here relies on some sort of magic handshake where whatever component actually creates the global will honor whatever alignment you can find in the bitcode?

That's pretty fragile -- usually in LLVM it's always correct to reduce the alignment attribute on such declarations, it just means we know less about what the real alignment is. Seems like some system here is (ab)using LLVM attributes in ways that break some of their core properties, or am I misunderstanding something?

Is this magic handshake documented anywhere?